### ISSUE TITLE
Using createHTOContigList() for single sequencing run returns empty list as output

### ISSUE THREAD

**Issue by ddressman91** [ncborcherding/scRepertoire#243]
Hello,

I have a Seurat object ("pbmc") that has 15 separate multiplexed sequencing runs integrated (106 individuals total).  Batch, antibody tag, and patient ID are in the metadata.  I loaded all the filtered_contig_annotations.csv files into R and added a merged_ids column to each of these, which supposedly matches the merged_ids variable in my Seurat object (merged_ids = [Batch name]_[Single-cell barcode]).  I then made the object "tcrlist", which is a list object combining all the objects from the csv files.  I should note that the Seurat object metadata does not currently have the HTO barcode sequences, but because each cell does have the patient ID in the metadata, those sequences could easily be added if necessary.  I first attempted to run

`contiglist <- createHTOContigList( contig = tcrlist, sc = pbmc, group.by = "merged_ids", multi.run = "Batch")`

This returned the following error:

`Error in `.rowNamesDF<-`(x, value = value) : 
  duplicate 'row.names' are not allowed
In addition: Warning message:
non-unique value when setting 'row.names': ‘NA’ `

I wasn't too suprised about this, since there are many single-cell barcodes that are duplicated even within a single sequencing run (due to the multiplexing in each run), not to mention how the multiple runs in the integrated Seurat object would compound this problem.  So I used SplitObject() in Seurat to split the object by the Batch variable back into the individual sequencing runs, created a Seurat object from the first batch, and then tried the following to attempt to demultiplex the TCR data for a single sequencing run:

`contiglist1 <- createHTOContigList(contig = tcrbatch1, sc = pbmcbatch1, group.by = "merged_ids")`

This completes without throwing errors, but contiglist1 is a list where each element is a unique member of merged_ids, with columns matching the columns of the TCR contig file, but no rows.  "pbmcbatch1" has 8464 unique members of merged_ids, and "tcrbatch1" has 4592, but only 50 of them overlap according to `length(intersect(tcrbatch1$merged_ids,pbmcbatch1$merged_ids))`.  But I tried looking at an element of contiglist1 that was one of the overlapping merged_ids, and it still had no rows.

Do you know what might be going on?  Thanks in advance for the help; I'm hoping to find a way to demultiplex the TCR data without having to use cellranger multi.

Dallin

**Comment by ncborcherding** [comment]
Hey Dallin, 

Thanks for reaching out and giving ```createHTOContigList()``` a thorough stress test. 

Honestly, when I wrote the function, it was a stop gap for emerging multiplex experiments and I did not envision a user using it for the over 10 individuals, let alone 100. There are some very simple assumptions that are leading to the issue with barcodes duplication. 

If you are willing - I would be happy to troubleshoot with you, as I do not have access to a real compendium of multiplexed experiments. Open to however you'd like to work it, just let me know.

Nick

**Comment by ddressman91** [comment]
I did some troubleshooting, and I think I figured it out.  I first added a "batch" column to each individual TCR contig file after reading it in, created a data frame of all the contig files ("tcrdf") with bind_rows, then edited the "barcode" column so that it was a composite of the batch and the single cell barcode (Batch_barcode being the syntax).  I also edited the row names of the Seurat object metadata to have the same Batch_barcode syntax.  I then ran `createHTOContigList(contig = tcrdf, sc = pbmc, group.by = "PatientID")` and that generated a list where each element was the contigs for an individual patient.  I didn't realize earlier that the group.by variable was supposed to be the sample-level variable in the Seurat metadata.  This ran much faster than my initial createHTOContigList() call from my earlier post.

Notably, this was done after the integration of the Seurat object (by "integration" on your Bioconductor page, I'm assuming you mean combination of multiple Seurat objects).  This integration was done by a colleague rather than by me, and however he did it, it didn't alter the barcodes.  So I had no need for the multi.run argument.
